src/compiler/scala/tools/nsc/settings/ScalaSettings.scala

```diff
@@ -134,6 +133,7 @@ trait ScalaSettings extends StandardScalaSettings with Warnings {
   val sourceReader       = StringSetting       ("-Xsource-reader", "classname", "Specify a custom method for reading source files.", "")
   val reporter           = StringSetting       ("-Xreporter", "classname", "Specify a custom subclass of FilteringReporter for compiler messages.", "scala.tools.nsc.reporters.ConsoleReporter")
   val source             = ScalaVersionSetting ("-Xsource", "version", "Treat compiler input as Scala source for the specified version, see scala/bug#8126.", initial = ScalaVersion("2.13"))
+                             .withPostSetHook(source => if (source.value >= version214 && source.value < version3) source.value = version3)
```

Comment:
did you consider / try adding `source.withDeprecationMessage`?

Reply:
No, I hadn't seen that.  Looks like if we want to use that we would need to adjust its usages, as it currently seems to be about deprecated setting key, not just some of their values:

```scala
runReporting.deprecationWarning(NoPosition, s.name + " is deprecated: " + s.deprecationMessage.get, "", "", "")
```

```scala
    def helpStr(s: Setting) = {
      val str    = layout(s.helpSyntax, s.helpDescription)
      val suffix = s.deprecationMessage match {
        case Some(msg) => "\n" + format("") + "      deprecated: " + msg
        case _         => ""
      }
      str + suffix
    }
```

Or maybe it's better to add a deprecation mechanism specific for setting values?

WDYT?  Preference?

Reply:
It's not fully accurate, but
```
-Xsource is deprecated: instead of -Xsource:2.14, use -Xsource:3
```
would be okish?

Reply:
Drop `-Xsource:3` for `-language:Scala3Compat`!